### ISSUE TITLE
IBM MQ AsyncAPI mocking/testing support

### ISSUE THREAD

**Issue by lbroudoux** [microcks/microcks#447]
As IBM MQ is a valid [AsyncAPI 2.1 Binding](https://github.com/asyncapi/bindings/tree/master/ibmmq), it may be a good candidate for integration within Microcks. Please vote for this feature if you need it!

@dalelane Is it something you're interested in ? ;-)

**Comment by dalelane** [comment]
Nice idea, thanks! I'll discuss it with some of my colleagues this week

**Comment by Evreke** [comment]
Any news on this?

**Comment by lbroudoux** [comment]
Hi @Evreke 
No news on this topic or people who voted for it recently. Please add your thumb-up reaction if it is something you'd be interested in! Also, if it's something you'd like to champion or contribute, I'd be happy to guide you!

**Comment by Evreke** [comment]
@lbroudoux 
I've never worked with IBM MQ, but I want to dive into microcks in order to be able to contribute or discuss it.

**Comment by lbroudoux** [comment]
Awesome! I'll try to guide you the best I can though. Here's the step I would go with:
1. Check the AsyncAPI Bindings for IBM MQ in order to see the different queue or interaction styles that way me defined on this protocol: https://github.com/asyncapi/bindings/tree/master/ibmmq
2. Add a new binding type in Microcks: https://github.com/microcks/microcks/blob/master/commons/model/src/main/java/io/github/microcks/domain/BindingType.java
3. Select the correct Java libraries and try to write a mock message producer manager. You can check the MQTT one for inspiration: https://github.com/microcks/microcks/blob/master/minions/async/src/main/java/io/github/microcks/minion/async/producer/MQTTProducerManager.java
4. The producer manager should then be bound to the global `ProducerManager` class: https://github.com/microcks/microcks/blob/master/minions/async/src/main/java/io/github/microcks/minion/async/producer/ProducerManager.java#L62
5. Write a test message consumption task for the tests. You can check the MQTT one for inspiration: https://github.com/microcks/microcks/blob/master/minions/async/src/main/java/io/github/microcks/minion/async/consumer/MQTTMessageConsumptionTask.java
6. Bind this message consumption task into global `TestManager` class: https://github.com/microcks/microcks/blob/master/minions/async/src/main/java/io/github/microcks/minion/async/AsyncAPITestManager.java#L313

Of course, this is long and probably hard work but that can be done through iteration. Usually, we create many issues to track this kind of Epic one! This is better for making progress and assuring you're on track. Do not hesitate to keep us posted on any of your concerns and progress!

Can I ask why this intesrest in contributing to Microcks? Do you or your company use it internally?

**Comment by Evreke** [comment]
We work in fintech and need to test our applications in an isolated environment. Therefore, we need to simulate a production environment. IBM MQ is the one that we cannot simulate now.

I'll try to check the code soon and start programming this weekend.

**Comment by lbroudoux** [comment]
Hey @Evreke 
I've created a bunch of smallest issues related to this Epic. I cannot assign you directly as you're not a maintainer but feel free to comment on those issues that you're working on that topic.

I didn't put a milestone yet - if possible it could be nice to make it land in `1.11.0` that is scheduled in ~2 months. There is no `1.11.0` branch at the moment ; I'll probably create it next week after the release of the `1.10.1`. In the meantime, you can start looking at the `1.10.x` branch - you'll just have to rebase on the new branch later on.

Cheers,

**Comment by Evreke** [comment]
There were urgent matters. I plan to continue by the weekend.

I had to teach my colleagues how to use microcks, and at the same time I was able to better understand the concept of headers.

**Comment by lbroudoux** [comment]
Thanks for keeping us in the loop!

**Comment by lbroudoux** [comment]
Hey! Just have a look at your commits above and it looks like you've made good progress! Anything I can do to help?

**Comment by lbroudoux** [comment]
Hi @Evreke 
I was wondering if you did have the chance to move forward on this IBM MQ integration?

**Comment by lbroudoux** [comment]
Hi @Evreke 
Just a gentle nudge... Anything new on this?

**Comment by Harsh4902** [comment]
@lbroudoux I would love to pick this up from where @Evreke left off and complete it. If you have no issues with that, I can move forward.

**Comment by lbroudoux** [comment]
> [@lbroudoux](https://github.com/lbroudoux) I would love to pick this up from where [@Evreke](https://github.com/Evreke) left off and complete it. If you have no issues with that, I can move forward.

💯 🎉 

More granular issues corresponding to the parts of this Epic have been created in #1342 #1343 #1344 and #1345

Do you have a sponsor or a testing use-case for this feature?

**Comment by Harsh4902** [comment]
> > [@lbroudoux](https://github.com/lbroudoux) I would love to pick this up from where [@Evreke](https://github.com/Evreke) left off and complete it. If you have no issues with that, I can move forward.
> 
> 💯 🎉
> 
> More granular issues corresponding to the parts of this Epic have been created in [#1342](https://github.com/microcks/microcks/issues/1342) [#1343](https://github.com/microcks/microcks/issues/1343) [#1344](https://github.com/microcks/microcks/issues/1344) and [#1345](https://github.com/microcks/microcks/issues/1345)
> 
> Do you have a sponsor or a testing use-case for this feature?

No, currently I don't have any sponsor or a testing use-case. I am still figuring out the scope of this feature to get better understanding.

**Comment by Evreke** [comment]
@lbroudoux I apologize for my disappearance. Unfortunately, I've been overloaded with my main job and other difficult circumstances.
I'd be happy if @Harsh4902 could use my modest resources to implement this feature.

I was stuck with IBM MQ headers. I had to dive in, and that's when life caught up with me.
I realized that most of the headers were parts that could have been ignored for the first step.

**Comment by lbroudoux** [comment]
> [@lbroudoux](https://github.com/lbroudoux) I apologize for my disappearance. Unfortunately, I've been overloaded with my main job and other difficult circumstances. I'd be happy if [@Harsh4902](https://github.com/Harsh4902) could use my modest resources to implement this feature.
> 
> I was stuck with IBM MQ headers. I had to dive in, and that's when life caught up with me. I realized that most of the headers were parts that could have been ignored for the first step.

No worries @Evreke and thanks for the heads-up. I hope things are going better for you now 🙏

**Comment by yada** [comment]
This issue has been pending for quite some time. If this functionality is important to you, consider this a call to action. To be introduced as a new binding in Microcks, we'd need to hear from adopters. If you're interested, please raise your hand, offer your help, commit to testing it, or even better, contribute to its implementation.
